### PR TITLE
Speed up toLocaleString ~2.5x by optimizing creation of Intl.DateTimeFormat instances

### PR DESCRIPTION
**Pull request by justingrant**
Port of https://github.com/js-temporal/temporal-polyfill/pull/12

### CODE REVIEW

**Comment by justingrant** [comment]
This PR wraps up a group of three DateTimeFormat optimization PRs. (The others were #1624 and #1629.)

Two weeks ago, demitasse tests took 46s-49s in CI. This PR only took 32s... 30% less time!

File: polyfill/package.json
Comment (Ms2ger) [inline polyfill/package.json:11]: Revert

File: polyfill/test/intl.mjs
Comment (Ms2ger) [inline polyfill/test/intl.mjs:1064]: Should this ever be called?

**Review by Ms2ger — APPROVED** [review]

File: polyfill/test/intl.mjs
Comment (justingrant) [reply polyfill/test/intl.mjs:1064]: Good catch. An early version of this PR used a JSON round-trip as a cheap way to do a deep copy of the input parameters, so this was previously useful. Now it's implemented differently so this `toJSON` test code will never run. I'll remove it, unless you think there's value in leaving it around in case some future implementation decides to use `JSON.stringify` on the inputs.

Comment (Ms2ger) [reply polyfill/test/intl.mjs:1064]: I'm happy to keep it and add `valueOf` as well, or just drop it.

File: polyfill/package.json
Comment (justingrant) [reply polyfill/package.json:11]: Will do. If OK with you, I'd like to leave it in the other repo. Especially as we start attracting new contributors over there, I'd like to instill awareness of perf regressions from everyone contributing code over there. Unless it hurts something?

File: polyfill/test/intl.mjs
Comment (justingrant) [reply polyfill/test/intl.mjs]: I just dropped it for simplicity. We already have to revisit this per https://github.com/js-temporal/temporal-polyfill/pull/12#discussion_r670048689 so any improvement in multi-call/proxy-calll tests could happen then.

I think we're good to merge. It's pretty late for me so I'm going to bed soon. If I can't stay up until the Test262 run finishes, I'll merge in the morning if you don't do it first.  Then I can go on vacation! :-)

**Review by Ms2ger — APPROVED** [review]

File: polyfill/package.json
Comment (Ms2ger) [reply polyfill/package.json]: I don't suppose it hurts, but I'm slowly removing these tests anyway, so their runtime doesn't matter much here. If you want to think about ways to speed up the 262 tests, that'd help me more :)

**Pull request merged**
